Optional component in properties

<!-- Please only include one feature request per issue, thank you! -->

**Describe the feature you'd like**
A clear and concise description of what you want to happen.

Currently a component can have children. However, with JSX you can also set a single component to a property, optionally:

~~~rust
#[derive(Clone, PartialEq, Properties)]
pub struct Props {
    pub children: ChildrenRenderer<Sections>,
    #[prop_or_default]
    pub sidebar: Option<Sidebar>,
}
~~~

**Is your feature request related to a problem? Please describe.** _(Optional)_

Currently you can do this is a list of children, but not with a single one.

You can achieve this currently only using `Html`:

~~~rust
#[derive(Clone, PartialEq, Properties)]
pub struct Props {
    pub children: ChildrenRenderer<Sections>,
    #[prop_or_default]
    pub sidebar: Html,
}
~~~

However, that doesn't ensure that the component is of a specific (required) type.

**Questionnaire**
<!-- Developing Yew is a community effort! -->
<!-- If you feel up to the challenge, please check one of the boxes below: -->
- [ ] I'm interested in implementing this myself but don't know where to start
- [ ] I would like to add this feature
- [X] I don't have time to add this right now, but maybe later

@siku2 Thanks for the explanation! That should definitely go into the documentation. However it solves my issue only partly, as I am trying to add an optional component.

When use your example, I must provide a sidebar. When I wrap `VChild<Sidebar>` with `Option<>`, I get the following error:

~~~
error[E0277]: `std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>` doesn't implement `std::fmt::Display`
  --> src/page.rs:57:19
   |
57 |                 { self.props.sidebar }
   |                   ^^^^ `std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>` cannot be formatted with the default formatter
   |
   = help: the trait `std::fmt::Display` is not implemented for `std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>`
   = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
   = note: required because of the requirements on the impl of `std::string::ToString` for `std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>`
   = note: required because of the requirements on the impl of `std::convert::From<std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>>` for `yew::virtual_dom::VNode`
   = note: required because of the requirements on the impl of `std::convert::Into<yew::virtual_dom::VNode>` for `std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>`
   = note: required because of the requirements on the impl of `std::convert::From<std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>>` for `yew::utils::NodeSeq<std::option::Option<yew::virtual_dom::VChild<pagesidebar::PageSidebar>>, yew::virtual_dom::VNode>`
   = note: required by `std::convert::From::from`
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

~~~

--- a/website/docs/concepts/components/children.md
+++ b/website/docs/concepts/components/children.md
@@ -92,7 +92,7 @@ pub enum Item {
     Secondary(VChild<Secondary>),
 }
 
-// Now, we implment `Into<Html>` so that yew knows how to render `Item`.
+// Now, we implement `Into<Html>` so that yew knows how to render `Item`.
 impl Into<Html> for Item {
     fn into(self) -> Html {
         match self {

--- a/website/docs/concepts/components/children.md
+++ b/website/docs/concepts/components/children.md
@@ -169,8 +169,9 @@ The page component can be called either with the sidebar or without:
 
     // Page with sidebar
     html! {
-        <Page>
+        <Page sidebar={html_nested!{
             <PageSideBar />
+        }}>
         </Page>
     }
 ```

--- a/website/docs/concepts/components/children.md
+++ b/website/docs/concepts/components/children.md
@@ -37,6 +37,7 @@ impl Component for List {
 
 ## Advanced usage
 
+### Typed children
 In cases where you want one type of component to be passed as children to your component,
 you can use `yew::html::ChildrenWithProps<T>`.
 
@@ -70,6 +71,7 @@ impl Component for ListProps {
 }
 ```
 
+### Enum typed children
 Of course, sometimes you might need to restrict the children to a few different
 components. In these cases, you have to get a little more hands-on with Yew.
 
@@ -124,3 +126,51 @@ impl Component for List {
 }
 ```
 
+### Optional typed child
+You can also have a single optional child component of a specific type too: 
+
+```rust
+use yew::prelude::*;
+use yew::virtual_dom::VChild;
+
+
+#[derive(Clone, Properties)]
+pub struct PageProps {
+    #[prop_or_default]
+    pub sidebar: Option<VChild<PageSideBar>>,
+}
+
+struct Page {
+    props: PageProps,
+}
+
+impl Component for Page {
+    type Properties = PageProps;
+    // ...
+
+    fn view(&self) -> Html {
+        html! {
+            <div class="page">
+                { self.props.sidebar.clone().map(Html::from).unwrap_or_default() }
+                // ... page content
+            </div>
+        }
+    }
+}
+```
+
+The page component can be called either with the sidebar or without: 
+
+```rust
+    // Page without sidebar
+    html! {
+        <Page />
+    }
+
+    // Page with sidebar
+    html! {
+        <Page>
+            <PageSideBar />
+        </Page>
+    }
+```

--- a/website/docs/concepts/components/children.md
+++ b/website/docs/concepts/components/children.md
@@ -169,9 +169,8 @@ The page component can be called either with the sidebar or without:
 
     // Page with sidebar
     html! {
-        <Page sidebar={html_nested!{
+        <Page sidebar=html_nested! {
             <PageSideBar />
-        }}>
-        </Page>
+        } />
     }
 ```

--- a/website/versioned_docs/version-0.18.0/concepts/components/children.md
+++ b/website/versioned_docs/version-0.18.0/concepts/components/children.md
@@ -92,7 +92,7 @@ pub enum Item {
     Secondary(VChild<Secondary>),
 }
 
-// Now, we implment `Into<Html>` so that yew knows how to render `Item`.
+// Now, we implement `Into<Html>` so that yew knows how to render `Item`.
 impl Into<Html> for Item {
     fn into(self) -> Html {
         match self {

--- a/website/versioned_docs/version-0.18.0/concepts/components/children.md
+++ b/website/versioned_docs/version-0.18.0/concepts/components/children.md
@@ -169,8 +169,9 @@ The page component can be called either with the sidebar or without:
 
     // Page with sidebar
     html! {
-        <Page>
+        <Page sidebar={html_nested!{
             <PageSideBar />
+        }}>
         </Page>
     }
 ```

--- a/website/versioned_docs/version-0.18.0/concepts/components/children.md
+++ b/website/versioned_docs/version-0.18.0/concepts/components/children.md
@@ -37,6 +37,7 @@ impl Component for List {
 
 ## Advanced usage
 
+### Typed children
 In cases where you want one type of component to be passed as children to your component,
 you can use `yew::html::ChildrenWithProps<T>`.
 
@@ -70,6 +71,7 @@ impl Component for ListProps {
 }
 ```
 
+### Enum typed children
 Of course, sometimes you might need to restrict the children to a few different
 components. In these cases, you have to get a little more hands-on with Yew.
 
@@ -124,3 +126,51 @@ impl Component for List {
 }
 ```
 
+### Optional typed child
+You can also have a single optional child component of a specific type too: 
+
+```rust
+use yew::prelude::*;
+use yew::virtual_dom::VChild;
+
+
+#[derive(Clone, Properties)]
+pub struct PageProps {
+    #[prop_or_default]
+    pub sidebar: Option<VChild<PageSideBar>>,
+}
+
+struct Page {
+    props: PageProps,
+}
+
+impl Component for Page {
+    type Properties = PageProps;
+    // ...
+
+    fn view(&self) -> Html {
+        html! {
+            <div class="page">
+                { self.props.sidebar.clone().map(Html::from).unwrap_or_default() }
+                // ... page content
+            </div>
+        }
+    }
+}
+```
+
+The page component can be called either with the sidebar or without: 
+
+```rust
+    // Page without sidebar
+    html! {
+        <Page />
+    }
+
+    // Page with sidebar
+    html! {
+        <Page>
+            <PageSideBar />
+        </Page>
+    }
+```

--- a/website/versioned_docs/version-0.18.0/concepts/components/children.md
+++ b/website/versioned_docs/version-0.18.0/concepts/components/children.md
@@ -169,9 +169,8 @@ The page component can be called either with the sidebar or without:
 
     // Page with sidebar
     html! {
-        <Page sidebar={html_nested!{
+        <Page sidebar=html_nested! {
             <PageSideBar />
-        }}>
-        </Page>
+        } />
     }
 ```

